Add docs structure to the repo and place for templates recipes

derberg opened the pull request:
- Add new docs folder and move all docs there
- Add new doc with templates recipes
- Made sure API doc generation still works

Open question: how we would like to structure the recipes doc

Fixes #196

jonaslagoni commented:
Should we provide some explanation to the snippets as well, or are we assuming people understand how to use Nunjucks in conjunction with the generator?

jonaslagoni commented on docs/templates-recipes.md:
```diff
@@ -0,0 +1,87 @@
+
+### Using Schema objects to generate types for a given programming language
+
+* Java example where `$$schema$$.java` file looks like this. You need to provide your own implementation of `toJavaType` filter:
```
Maybe we should add the filter for completeness, something like the following (cannot figure out how to nest codeblocks, so can't make it as a suggestion 🙄): 

* Java example for `$$schema$$.java` file with provided `toJavaType` filter:

With the `toJavaType` filter for convenience: 
```js
...
  Nunjucks.addFilter('toJavaType', jsonSchemaType => {
    switch (jsonSchemaType.toLowerCase()) {
      case 'string':
        return 'String';
      case 'integer':
        return  'Integer'
      case 'number':
        return 'Double';
      case 'boolean':
        return 'boolean';
    }
  });
...
``

derberg commented:
@jonaslagoni hey man, as we discussed on slack and per your comment, I extended the PR, have a look

jonaslagoni commented on docs/authoring.md:
```diff
@@ -11,6 +11,9 @@ The AsyncAPI generator has been built with extensibility in mind. The package us
 1. Templates may contain "filters" or helper functions. They must be stored in the `.filters` directory under the template directory. [Read more about filters](#filters).
 1. Templates may have a configuration file. It must be stored in the template directory and its name must be `.tp-config.json`. [Read more about the configuration file](#configuration-file).
 1. There are params with special meaning. [Read more about special params](#special-params).
+1. Default parameters that you can use in your templates, except the ones that you defined on your own in the config,  are:
```
```suggestion
1. The default variables you have access to in any the template file are the following:
```
Think we have to use `variable` instead of parameters, not to confuse it with the configuration parameters.

jonaslagoni commented on docs/authoring.md:
```diff
@@ -11,6 +11,9 @@ The AsyncAPI generator has been built with extensibility in mind. The package us
+1. Default parameters that you can use in your templates, except the ones that you defined on your own in the config,  are:
+   - `asyncapi` that is a parsed spec file object. Read the [API](https://github.com/asyncapi/parser-js/blob/master/API.md#AsyncAPIDocument) of the Parser to understand to what structure you have access in this parameter.
+   - `originalAsyncAPI` that is an original spec file before it is parsed. 
 
```
```suggestion
1. When using [file templating](https://github.com/asyncapi/generator#file-templating) you have access to the following extra variables:
   - When using `$$channel$$` you have access to two variables `channel` and `channelName`. Where `channel` contains the current channel being rendered.
   - When using `$$message$$` you have access to two variables `message` and `messageName`. Where `message` contains the current message being rendered.
   - When using `$$schema$$` you have access to two variables `schema` and `schemaName`. Where `schema` contains the current schema being rendered.
   - When using `$$parameter$$` you have access to two variables `parameter` and `parameterName`. Where `parameter` contains the current parameter being rendered.
   - When using `$$securityScheme$$` you have access to two variables `securityScheme` and `securitySchemeName`. Where `securityScheme` contains the current security scheme being rendered.
```
Think we should add the file template variables we have access to when using that feature 😄

jonaslagoni commented on docs/authoring.md:
```diff
@@ -102,4 +105,4 @@ There are some template parameters that have a special meaning:
 |Name|Description|
 |---|---|
-|`server`| It is used to let the template know which server we want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. If your template need your users to specify a server, use this param.
+|`server`| It is used to let the template know which server we want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. If your template need your users to specify a server, use this param.
```
```suggestion
|`server`| It is used to let the template know which server we want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. If your template need your users to specify a server, use this parameter.
```

jonaslagoni commented on docs/templates-recipes.md:
```diff
@@ -0,0 +1,110 @@
+        return 'Double';
+        case 'boolean':
+        return 'boolean';
+    }
```
```suggestion
        switch (jsonSchemaType.toLowerCase()) {
            case 'string':
                return 'String';
            case 'integer':
                return  'Integer'
            case 'number':
                return 'Double';
            case 'boolean':
                return 'boolean';
        }
```
Indentation was missing 🤔

derberg replied:
you sure, lgtm in the preview

jonaslagoni replied:
@derberg this is still the case if you preview the file 😄 Or at least for me 🤔 ?

derberg replied:
@jonaslagoni it all depends what you expect 😆 I see this and I expect this 😆 
<img width="908" alt="Screenshot 2020-02-28 at 21 22 34" src="https://user-images.githubusercontent.com/6995927/75584666-860c2000-5a70-11ea-8b9e-604f108aa822.png">

jonaslagoni replied:
I just wanted the switch statement to have indentation as well :smile:

derberg commented:
@jonaslagoni thanks for the input about file templating. I cleaned it up and removed it entirely from readme as its plays was fully in the authoring doc. I also applied your suggestions and made sure `params` is not used at all in authoring doc.

There is just one conversation unresolved, I don't know whay you mean by the missing  indentation as it lgtm
